### ISSUE TITLE
Resize image before upload

### ISSUE THREAD

**Issue by riaan53** [vsivsi/meteor-file-collection#29]
Hi,

Im trying to resize an image before uploading. I would like to do this with resumable.js if possible - but it only accepts a File. On chrome and Firefox I can use `new File` and create a file from a canvas blob that can be fed into resumable. But this is not a cross browser solution. Do you know of a way to do this? If not with resumable.js how do I upload without it?

See ticket on resumable.js github as well. https://github.com/23/resumable.js/issues/202

Regards,
Riaan

**Comment by vsivsi** [comment]
@riaan53  Hi, I'm just recently back from vacation. Thanks for this questions and for the other work you've done on other topics. I'll starting working on these tomorrow.  Thanks for your patience!

**Comment by riaan53** [comment]
Hi @vsivsi ,

No problem at all. Hope you had a great vacation!

I want to use file-collection to upload fotos/video/files from cordova as well using http://plugins.cordova.io/#/package/org.apache.cordova.camera. These days fotos are easily over 4mb so to resize/modify/compress them before upload is a must. The camera package can return the data in base64, fileURI, nativeURI. Its the easiest to use base64 to resize the image and to correct the orientation before upload and it only takes a few milliseconds using the html canvas.  Videos on the other hand will have to be resized on the server with ffmpeg streams (https://www.npmjs.com/package/fluent-ffmpeg), because it takes way to long to. For videos to work on all (most) browsers you have to encode them to mp4, webm and ogg formats. It will work great with way you are converting images in https://github.com/vsivsi/meteor-job-collection, maybe offloading the encoding to a new vm.

So basically for now I need to be able to not only feed a file into resumable.js. base64 or a blob will help a lot but without a rewrite to resumable it doest seem possible. The other option is to convert a blob to a html file api object - dont know if its possible but the File class objects inherits from the BLOB class. And then lastly is not to use rsumable.js, but I need a robust/resumable solution so will have to integrate something else into file-collection.

Any ideas?

Thank you for the awesome package!! 

Regards,
Riaan

**Comment by vsivsi** [comment]
Hi Riaan, You don't have to use resumable.js to support uploads with file-collection if it doesn't meet your needs.  I just packaged it in for convenience.   

The easiest alternative way to upload is to just define your own HTTP POST or PUT handler as described in the docs: https://github.com/vsivsi/meteor-file-collection#configuring-http-methods  Of course what you lose with this approach is the resumable.js chunked upload and auto-retry / resume goodness.

Then on the client, you can use the Meteor HTTP calls to generate POST or PUT requests.  http://docs.meteor.com/#/full/http  I'm not a Cordova expert and haven't written a lot of code to manipulate File Objects using js on the client-side, so I'm not sure if the Meteor HTTP call support is flexible enough to get your data from a File Object into the Meteor HTTP call in a way that will work with file-collection.  file-collection implements server side file POST and PUT handling in a pretty vanilla/standard way as can be seen in the example `curl` commands: https://github.com/vsivsi/meteor-file-collection#http-request-behaviors

If Meteor's HTTP support is inadequate, you are of course free to use any other client side library you would like generate those calls (e.g. jQuery AJAX calls, or even good old XMLHttpRequest).

I hope that helps.

**Comment by riaan53** [comment]
Hi @vsivsi ,

I have figured it out. Pretty simple at the end :)

Basically resumable.js will accept a blob with a name.

So the steps go like this:
1. Load image onto canvas and manipulate like you want.
2. Create base64 from canvas `canvas.toDataURL("image/jpeg");`
3. Create blob from base64 data.
4. Add a name to blob object
5. `resumable.addFile(myBlob);`

```
function dataURItoBlob(dataURI) {
    var byteString = atob(dataURI.split(',')[1]);
    var mimeString = dataURI.split(",")[0].split(":")[1].split(";")[0];
    var ab = new ArrayBuffer(byteString.length);
    var ia = new Uint8Array(ab);
    for (var i = 0; i < byteString.length; i++) {
        ia[i] = byteString.charCodeAt(i);
    }
    return new Blob([ab], { type: mimeString });
}

var myBlob = dataURItoBlob(canvas.toDataURL("image/jpeg"));
myBlob.name = "image.jpeg";
Media.resumable.addFile(myBlob); // My file collection name is Media
```

So with this method I can use resumable.js for Cordova uploads as well! There are some minor changes that I have made file-collection to support resumable.js uploads on Cordova (Request url and CORS). I will submit a PR once you accept the other one.

Regards,
Riaan

**Comment by vsivsi** [comment]
Great, thanks for sharing the solution you worked out to this issue.

**Comment by miri-am** [comment]
@riaan53. Can you provide more details on how you got his working for Cordova? It is working fine for me for browsers, but on Cordova empty files are uploaded.
